beacon-chain/forkchoice/protoarray/store.go

```diff
@@ -195,6 +196,56 @@ func (f *ForkChoice) AncestorRoot(ctx context.Context, root [32]byte, slot types
 	return f.store.nodes[i].root[:], nil
 }
 
+// CommonAncestorRoot returns the common ancestor root between the two block roots r1 and r2.
+func (f *ForkChoice) CommonAncestorRoot(ctx context.Context, r1 [32]byte, r2 [32]byte) ([32]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "protoArray.CommonAncestorRoot")
+	defer span.End()
+
+	// Do nothing if the two input roots are the same.
+	if r1 == r2 {
+		return r1, nil
+	}
+
+	f.store.nodesLock.RLock()
+	defer f.store.nodesLock.RUnlock()
+
+	i, ok := f.store.nodesIndices[r1]
+	if !ok || i >= uint64(len(f.store.nodes)) {
+		return [32]byte{}, errInvalidNodeIndex
+	}
+	n1 := f.store.nodes[i]
+
+	i, ok = f.store.nodesIndices[r2]
+	if !ok || i >= uint64(len(f.store.nodes)) {
+		return [32]byte{}, errInvalidNodeIndex
+	}
+	n2 := f.store.nodes[i]
+
+	for {
+		if ctx.Err() != nil {
+			return [32]byte{}, ctx.Err()
+		}
+		if n1.slot > n2.slot {
+			i = n1.parent
+			// Reaches the end of the tree and unable to find common ancestor.
+			if i >= uint64(len(f.store.nodes)) {
+				return [32]byte{}, forkchoice.ErrUnknownCommonAncestor
+			}
+			n1 = f.store.nodes[i]
+		} else {
+			i = n2.parent
+			// Reaches the end of the tree and unable to find common ancestor.
+			if i >= uint64(len(f.store.nodes)) {
+				return [32]byte{}, forkchoice.ErrUnknownCommonAncestor
+			}
+			n2 = f.store.nodes[i]
+		}
+		if n1.root == n2.root {
+			return n1.root, nil
+		}
```

Comment:
I think you can avoid the extra slot comparison by using:
```suggestion
	i1, ok := f.store.nodesIndices[r1]
	if !ok || i1 >= uint64(len(f.store.nodes)) {
		return [32]byte{}, errInvalidNodeIndex
	}

	i2, ok = f.store.nodesIndices[r2]
	if !ok || i2 >= uint64(len(f.store.nodes)) {
		return [32]byte{}, errInvalidNodeIndex
	}

	for {
		if ctx.Err() != nil {
			return [32]byte{}, ctx.Err()
		}
		if i1 > i2 {
			n1 := f.store.nodes[i1]
			i1 = n1.parent
			// Reaches the end of the tree and unable to find common ancestor.
			if i1 >= uint64(len(f.store.nodes)) {
				return [32]byte{}, forkchoice.ErrUnknownCommonAncestor
			}
		} else {
			n2 := f.store.nodes[i2]
			i2 = n2.parent
			// Reaches the end of the tree and unable to find common ancestor.
			if i2 >= uint64(len(f.store.nodes)) {
				return [32]byte{}, forkchoice.ErrUnknownCommonAncestor
			}
		}
		if i1 == i2 {
			n1 := f.store.nodes[i1]
			return n1.root, nil
		}
	}
```